### ISSUE TITLE
imperativeStatus and validationModes

### ISSUE THREAD

**Issue by nickevansuk** [openactive/data-models#41]
## Use case

Representing different property imperative configurations across validationModes

## Proposal

For a particular subset of `validationMode` (e.g. `C1Request`, `C2Request`) a specific `imperativeConfiguration` should be applied.

If no matching `validationMode` is found for the current mode, the default values for the `imperativeConfiguration` from the model root should be used.

If a matching `validationMode` is found for the current mode, then the `imperativeConfiguration` it references should be used and the values of `requiredFields`, `recommendedFields`, `shallNotInclude`  in the model root should be completely ignored

`shallNotInclude` should be added just within `imperativeConfiguration`, to represent properties that are not expected to be found in a particular `imperativeConfiguration`, but are still in the spec.

## Example

```
{
   "type":"Lease",
   "subClassOf": "https://schema.org/Thing",
   "validationMode":{
     "C1Request": "request",
     "C2Request": "request"
   }
   "imperativeConfiguration":{
      "request":{
         "hasId":false,
         "requiredFields":[
            "type",
            "leaseExpires"
         ],
         "recommendedFields":[

         ],
         "shallNotInclude":[

         ],
         "requiredOptions":[

         ]
      }
   },
```

### Rules
- New rule to recognise `shallNotInclude` and produce error messages as appropriate (note that `shallNotInclude` does not exist in the default mode, and only exists in `imperativeConfiguration`)
- Update to the existing rules for "Required" and "Recommended" to respect the input `validationMode` and resolve/use the relevant `imperativeConfiguration` instead of the default.

### Data model tests
- Data model tests to ensure that all fields featured in the `imperativeConfiguration` within `requiredFields`, `recommendedFields`, `shallNotInclude`, and `requiredOptions` exist in `inSpec` within the same model
- Data model tests to ensure that all `validationMode`s specified are valid (from the enum we've put somewhere)
- Data model tests to ensure that all `validationMode`s reference valid `imperativeConfiguration`s within the same model

**Comment by henryaddison** [comment]
@nickevansuk I have made a start on some of the rules and model specs for this as you can see in the two PRs linked above. Third one to follow soon (work done just need to push it when I can access it again). This probably marks about the end of the initial timebox for working on the model validator library's rules (I might be able to squeeze in moving the source of truth for available validation modes from the validator library to the data-models repo).

**Comment by henryaddison** [comment]
@nickevansuk I am getting through the work for this issue and so thinking a bit how these changes will be deployed.

Adding validationModeGroup to the metadata in `data-models` doesn't break anything but the validator site will come to rely on it. My suggestion is therefore to make a new minor version of `data-models` (1.6.0 from 1.5.6 I think) so the changes to the validator site can specify only later versions.

However, the changes to the examples/samples is more complex as rather than a pure addition it is a change to `data-models` which is not backwards compatible with the validator site code (before the validator site expects examples to be a flat list of objects, after it is groups of lists). I assume this is not a case of creating a new version of the spec in data models but instead should be a change of version of the `@openactive/data-models` npm pacakge, which I think strictly should be a new major version (2.0) to indicate the breaking change.

Basically do you agree with me updating the npm package's version to 2.0.0 in #46 ?

**Comment by nickevansuk** [comment]
yep updating the models to 2.0 is fine, we're doing fairly major tooling work across the board so makes sense to me

**Comment by thtg88** [comment]
If I understand this correctly does this mean we will also have to bump major in [`package.json`](https://github.com/openactive/models-lib/blob/master/package.json#L21)?

**Comment by henryaddison** [comment]
@thtg88 Other than the change to the Barcode.json file name none of the changes I'm making effect the models-lib (yet) and even the file rename doesn't seem to be an issue for the code generator either so you shouldn't HAVE to but it may probably be wise to update the version in models-lib once it's released to avoid failing behind.

**Comment by henryaddison** [comment]
I have noticed that the Rules in data-model-validator lib are quite highly coupled to a version of the spec in data-models lib due to the way rules are target fields, models and validation modes. This is unavoidable - for new versions of the spec there will need to be some new rules and some old rules will no longer apply. However, the way it is currently implemented means the rules are at risk of accidentally being ignored due to types in the strings used by targetFields, targetModels and targetValidationModes.

At some point I guess it will be necessary to have per-version collections of the rules to apply during validation and therefore more explicitly include version when constructing a new include of a Rule class (maybe this is already possible since constructor takes an OptionsHelper object).

**Comment by henryaddison** [comment]
@nickevansuk & @thill-odi,

With the merging of the changes to the validator site above (https://github.com/openactive/data-model-validator-site/pull/56), I believe that completes the work for this issue (bar possibly our comments about validating validationModes, targetModels and targetFields - this has been added to bottom of spreadsheet and perhaps belongs in it's own issue).

Work is merged into master but I will leave to you to confirm all is good and to deploy the updates to the production site (not least because I don't know how it's hosted!).

**Comment by nickevansuk** [comment]
@henryaddison still getting an error at https://validator-staging.openactive.io/ ? Assume it works on your machine?

**Comment by nickevansuk** [comment]
In fact @henryaddison here's the logs from Heroku:

```
2019-11-07T10:39:00.902400+00:00 heroku[web.1]: State changed from crashed to starting
2019-11-07T10:39:05.148534+00:00 heroku[web.1]: Starting process with command `npm start`
2019-11-07T10:39:07.357507+00:00 app[web.1]: 
2019-11-07T10:39:07.357538+00:00 app[web.1]: > data-model-validator-site@0.1.0 start /app
2019-11-07T10:39:07.357539+00:00 app[web.1]: > node -r dotenv/config dist/server/index.js
2019-11-07T10:39:07.357540+00:00 app[web.1]: 
2019-11-07T10:39:08.256822+00:00 app[web.1]: /app/dist/server/server.js:81
2019-11-07T10:39:08.256875+00:00 app[web.1]:         var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(req, res) {
2019-11-07T10:39:08.256877+00:00 app[web.1]:                                                    ^
2019-11-07T10:39:08.256879+00:00 app[web.1]: 
2019-11-07T10:39:08.256880+00:00 app[web.1]: ReferenceError: regeneratorRuntime is not defined
2019-11-07T10:39:08.256883+00:00 app[web.1]:     at /app/dist/server/server.js:81:52
2019-11-07T10:39:08.256884+00:00 app[web.1]:     at Function.createServer (/app/dist/server/server.js:208:8)
2019-11-07T10:39:08.256885+00:00 app[web.1]:     at Object.<anonymous> (/app/dist/server/index.js:13:18)
2019-11-07T10:39:08.256887+00:00 app[web.1]:     at Module._compile (module.js:652:30)
2019-11-07T10:39:08.256888+00:00 app[web.1]:     at Object.Module._extensions..js (module.js:663:10)
2019-11-07T10:39:08.256889+00:00 app[web.1]:     at Module.load (module.js:565:32)
2019-11-07T10:39:08.256891+00:00 app[web.1]:     at tryModuleLoad (module.js:505:12)
2019-11-07T10:39:08.256892+00:00 app[web.1]:     at Function.Module._load (module.js:497:3)
2019-11-07T10:39:08.256893+00:00 app[web.1]:     at Function.Module.runMain (module.js:693:10)
2019-11-07T10:39:08.256894+00:00 app[web.1]:     at startup (bootstrap_node.js:191:16)
2019-11-07T10:39:08.264065+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2019-11-07T10:39:08.264360+00:00 app[web.1]: npm ERR! errno 1
2019-11-07T10:39:08.265502+00:00 app[web.1]: npm ERR! data-model-validator-site@0.1.0 start: `node -r dotenv/config dist/server/index.js`
2019-11-07T10:39:08.265632+00:00 app[web.1]: npm ERR! Exit status 1
2019-11-07T10:39:08.265867+00:00 app[web.1]: npm ERR! 
2019-11-07T10:39:08.266253+00:00 app[web.1]: npm ERR! Failed at the data-model-validator-site@0.1.0 start script.
2019-11-07T10:39:08.266467+00:00 app[web.1]: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
2019-11-07T10:39:09.689298+00:00 heroku[web.1]: State changed from starting to crashed
2019-11-07T10:39:09.626363+00:00 app[web.1]: 
2019-11-07T10:39:09.626549+00:00 app[web.1]: npm ERR! A complete log of this run can be found in:
2019-11-07T10:39:09.626645+00:00 app[web.1]: npm ERR!     /app/.npm/_logs/2019-11-07T10_39_08_267Z-debug.log
2019-11-07T10:39:09.675431+00:00 heroku[web.1]: Process exited with status 1
```

**Comment by henryaddison** [comment]
Ah I'd foolishly assumed the issues with the staging site were unrelated. I will take a look.

**Comment by henryaddison** [comment]
On plus side can re-create on my machine with `npm start` rather than running the dev server.
